File: rig4j/src/main/java/com/alflabs/rig4/Timing.java
package com.alflabs.rig4;

import com.alflabs.annotations.NonNull;
import com.alflabs.utils.IClock;
import com.alflabs.utils.ILogger;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.Map;
import java.util.TreeMap;

@Singleton
public class Timing {
    private final IClock mClock;
    private final ILogger mLogger;
    private Map<String, TimeAccumulator> mMap = new TreeMap<>();

    @Inject
    public Timing(IClock clock, ILogger logger) {
        mClock = clock;
        mLogger = logger;
    }

    public TimeAccumulator get(@NonNull String name) {
        TimeAccumulator a = mMap.get(name);
        if (a == null) {
            a = new TimeAccumulator(name);
            mMap.put(name, a);
        }
        return a;
    }

    public void printToLog() {
        for (Map.Entry<String, TimeAccumulator> entry : mMap.entrySet()) {
            mLogger.d("Timing",
                    String.format("%s = %.3f s", entry.getKey(),
                            entry.getValue().getAccumulator() / 1000.0f));
        }
    }

    public class TimeAccumulator {
        private final String mName;
        private long mStart;
        private long mAccumulator;

        public TimeAccumulator(@NonNull String name) {
            mName = name;
        }

        public TimeAccumulator start() {
            mStart = mClock.elapsedRealtime();
            return this;
        }

        public void end() {
            if (mStart > 0) {
                long delay = mClock.elapsedRealtime() - mStart;
                mAccumulator += delay;
                mStart = 0;
            }
        }

        public String getName() {
            return mName;
        }

        public long getAccumulator() {
            return mAccumulator;
        }
    }
}

File: rig4k/src/main/kotlin/com/alflabs/rig4k/common/Timing.kt
package com.alflabs.rig4k.common

import com.alflabs.utils.IClock
import com.alflabs.utils.ILogger
import java.util.TreeMap
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class Timing @Inject constructor(
    private val clock: IClock,
    private val logger: ILogger
) {
    private val map: MutableMap<String, TimeAccumulator> = TreeMap()

    fun get(name: String): TimeAccumulator {
        var a = map[name]
        if (a == null) {
            a = TimeAccumulator(name)
            map[name] = a
        }
        return a
    }

    fun printToLog() {
        for ((key, value) in map) {
            logger.d(
                "Timing", String.format(
                    "%s = %.3f s", key,
                    value.accumulator / 1000.0f
                )
            )
        }
    }

    inner class TimeAccumulator(val name: String) {
        private var start: Long = 0
        var accumulator: Long = 0
            private set

        fun start(): TimeAccumulator {
            start = clock.elapsedRealtime()
            return this
        }

        fun end() {
            if (start > 0) {
                val delay = clock.elapsedRealtime() - start
                accumulator += delay
                start = 0
            }
        }
    }
}
